### PR TITLE
Added support for non-string JOSE headers.

### PR DESCRIPTION
**Pull request by russjones**
**Description**

Added `UnmarshalJSON` function that converts float64 values into strings when unmarshalling JOSE headers. This is needed because IBM Cloud AppID IdP returns non-string values for some header fields.

**Related Issues**

https://github.com/gravitational/teleport/issues/6717

### CODE REVIEW

**Comment by awly** [comment]
@russjones can we update `github.com/coreos/go-oidc` to `v3` instead?
In `v3` they use https://github.com/square/go-jose for parsing tokens and their header parsing looks like it supports non-string values.

Try `go get github.com/coreos/go-oidc@v3` in `teleport` and then update import paths and usage (the API changed).

FYI, it looks like JWT parsing logic was moved to the `square` repo back in 2016: https://github.com/coreos/go-oidc/issues/74

**Comment by russjones** [comment]
@awly We could but I think we maintain other patches in our fork for weird IdP edge cases. Might be a larger unit of work, not even sure if we can test the changes since we probably never logged who those weird patches were made for.

File: jose/jose.go [hunk 1]
```diff
@@ -66,6 +66,33 @@ const (
...
+		return err
+	}
+
+	out := make(map[string]string)
+	for k, v := range s {
+		switch v.(type) {
+		case float64:
+			out[k] = fmt.Sprint(v.(float64))
+		case string:
+			out[k] = v.(string)
```
Comment (awly) [inline jose/jose.go:86]:
```suggestion
		switch vv := v.(type) {
		case float64:
			out[k] = fmt.Sprint(vv)
		case bool:
			out[k] = fmt.Sprint(vv)
		case string:
			out[k] = vv
```

**Review by awly — APPROVED** [review]
OK, at some point all the fixes (and more) will have already been fixed upstream and we're stuck maintaining our own fork.

image

**Comment by russjones** [comment]
@awly I agree, I do want to get off this library, we just have to make time for it at some point.
